addons/account/models/account_move.py

```diff
@@ -890,6 +896,19 @@ def _compute_currency_id(self):
             )
             invoice.currency_id = currency
 
+    @api.depends('currency_id', 'company_currency_id', 'company_id', 'invoice_date')
+    def _compute_invoice_currency_rate(self):
+        for move in self:
+            if move.currency_id and move.is_invoice(include_receipts=True):
+                move.invoice_currency_rate = self.env['res.currency']._get_conversion_rate(
+                    from_currency=move.company_currency_id,
+                    to_currency=move.currency_id,
+                    company=move.company_id,
+                    date=move.invoice_date or move.date or fields.Date.context_today(move),
```

Comment:
I am not sure I understand your concern.
Do you mean when the field is stored or the conversion date we use?
For l10n_ar the rate is (currently) stored during post to get the most up to date rate (since the field is not actually used for any computations anyway).
But it fetches the rate from the invoice_date. 
https://github.com/odoo/odoo/blob/4fcce49bee3fbbbda0f8f2e8ade524ef5f44dc07/addons/l10n_ar/models/account_move.py#L164-L176
Since it is done at post there is an invoice_date. But now we also need a currency rate before an invoice_date is set since we need it to determine the currency_rate of all the lines. So it falls back to the other dates.

Reply:
> So it falls back to the other dates.

Yes, this is my concern. The meaning of the field is different; I don't know if it's better or worse.

Reply:
I removed the fallback to `date`.
I think there is no point in this fallback since we only store the rate for invoices.

I looked into it a bit and I think it should be fine now.

I think the meaning is basically the same (just extended).
The AR field just has no value before posting and thus no "meaning" before an invoice date has been set.
If we consider only the time after an invoice date has been set the 2 calls are the same.
Here we simply extend the meaning to the time before an invoice date has been set.
For this we introduce a fallback to "today".
This is the correct choice imo since when no invoice date is set on post we use "today" as invoice_date [^1].

So the only thing left that could go wrong is going from having no invoice_date to having an invoice date.
In this case we (may) recompute the lines (when the rate or "today" has changed).
This keeps the exchange rate consistent with the invoice_date.
I checked with labe that it is acceptable.


[^1]: https://github.com/odoo/odoo/blob/6b6c56b67384183df2456a0055689b0294adf2b9/addons/account/models/account_move.py#L4044